syntax = "proto3";

package books_database;

service BooksDatabase {
  rpc Read(ReadRequest) returns (ReadResponse);
  rpc Write(WriteRequest) returns (WriteResponse);
  rpc DecrementStock(StockRequest) returns (StockResponse); // Bonus atomic operation
}

message ReadRequest {
  string title = 1;
}

message ReadResponse {
  int32 stock = 1;
}

message WriteRequest {
  string title = 1;
  int32 new_stock = 2;
}

message WriteResponse {
  bool success = 1;
}

message StockRequest {
  string title = 1;
  int32 quantity = 2;
}

message StockResponse {
  bool success = 1;
  int32 remaining = 2;
}
